lib/aiken/rationals.ak

```diff
@@ -0,0 +1,295 @@
+// Rationals
+
+use aiken/math
+
+// Useful for using Rationals in datums/redeemers which cannot have opaque types
+pub type UncheckedRational {
+  numerator: Int,
+  denominator: Int,
+}
+
+// Opaque type used to ensure the sign of the Rational is managed strictly
+// in the numerator
+pub opaque type Rational {
+  numerator: Int,
+  denominator: Int,
+}
+
+// Converts an UncheckedRational to a Rational
+pub fn check_rational(num: UncheckedRational) -> Rational {
+  let UncheckedRational { numerator, denominator } =
+    num
+  div_int(numerator, denominator)
+}
+
+// Multiplication
+pub fn mul(a: Rational, b: Rational) -> Rational {
+  Rational {
+    numerator: a.numerator * b.numerator,
+    denominator: a.denominator * b.denominator,
+  }
+}
+
+test mul_1() {
+  let a =
+    div_int(2, 3) |> mul(div_int(3, 4))
+  a == Rational { numerator: 6, denominator: 12 }
+}
+
+test mul_2() {
+  let a =
+    div_int(-2, 3) |> mul(div_int(-3, 4))
+  a == Rational { numerator: 6, denominator: 12 }
+}
+
+// Division
+pub fn div(a: Rational, b: Rational) -> Rational {
+  recip(b) |> mul(a)
+}
+
+test div_1() {
+  let a =
+    div_int(2, 3) |> div(div_int(3, 4))
+  a == Rational { numerator: 8, denominator: 9 }
+}
+
+test div_2() {
+  let a =
+    div_int(2, 3) |> div(div_int(-3, 4))
+  a == Rational { numerator: -8, denominator: 9 }
+}
+
+// Create a new Rational
+pub fn div_int(numerator: Int, denominator: Int) -> Rational {
+  if denominator < 0 {
+    Rational { numerator: -numerator, denominator: -denominator }
+  } else {
+    Rational { numerator, denominator }
+  }
+}
```

Comment:
This name seems misleading to me. The point is to create a fraction so I would prefer the naming convention used in PlutusTx.Ratio of `unsafeRatio` and `ratio`. This `div_int` function fits neither case though since it allows creating a fraction with a zero denominator.

Reply:
We should check against zero denominator. div_int expands to divide_integer, which is obviously what we're doing- don't see problem with name. It is not meant to be unsafe ratio - we don't intend to have any unsafe ratio, which is why we use the opaques.

Reply:
> div_int expands to divide_integer, which is obviously what we're doing- don't see problem with name. 

I guess this is semantics but in my mind, creating a fraction is **not** the same as dividing an `Int`. They are functionally different: one yields a fraction while the other yields another `Int`.

Having both an `unsafe` and a safe one allows for more generalization. The `unsafe` is meant to crash the script if the denominator is zero while the safe version allows for recovering when a denominator is zero. Both have there place. Using opaques does not change these use cases.

Reply:
Unsafe in terms of non-pure/failing I agree with, I thought you meant unsafe as in never checking the invariant is maintained.